dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeGenericDataModelMapper.cs

```diff
@@ -0,0 +1,101 @@
+        propertyReader.VerifyDataProperties(PineconeVectorStoreRecordFieldMapping.s_supportedDataTypes, PineconeVectorStoreRecordFieldMapping.s_supportedEnumerableDataElementTypes);
+        propertyReader.VerifyVectorProperties(PineconeVectorStoreRecordFieldMapping.s_supportedVectorTypes);
+
+        // Assign.
+        this._propertyReader = propertyReader;
+    }
+
+    /// <inheritdoc />
+    public Vector MapFromDataToStorageModel(VectorStoreGenericDataModel<string> dataModel)
+    {
+        var metadata = new MetadataMap();
+
+        // Map data properties.
+        foreach (var dataProperty in this._propertyReader.DataProperties)
+        {
+            if (dataModel.Data.TryGetValue(dataProperty.DataModelPropertyName, out var propertyValue))
+            {
+                var propertyStorageName = this._propertyReader.GetStoragePropertyName(dataProperty.DataModelPropertyName);
+                metadata[propertyStorageName] = propertyValue == null ?
+                    new MetadataValue() :
+                    PineconeVectorStoreRecordFieldMapping.ConvertToMetadataValue(propertyValue);
+            }
+        }
+
+        // Map vector property.
+        if (dataModel.Vectors.Count != 1)
+        {
+            throw new VectorStoreRecordMappingException($"Exactly one vector is supported by the Pinecone connector, but the provided data model contains {dataModel.Vectors.Count}.");
+        }
+
+        if (!dataModel.Vectors.TryGetValue(this._propertyReader.FirstVectorPropertyName!, out var valuesObject) || valuesObject is not ReadOnlyMemory<float> values)
+        {
+            throw new VectorStoreRecordMappingException($"Vector property '{this._propertyReader.FirstVectorPropertyName}' on provided record of type {nameof(VectorStoreGenericDataModel<string>)} must be of type ReadOnlyMemory<float> and not null.");
+        }
+
+        // TODO: what about sparse values?
+        var result = new Vector
+        {
+            Id = (string)dataModel.Key,
+            Values = values.ToArray(),
+            Metadata = metadata,
+            SparseValues = null
+        };
+
+        return result;
+    }
+
+    /// <inheritdoc />
+    public VectorStoreGenericDataModel<string> MapFromStorageToDataModel(Vector storageModel, StorageToDataModelMapperOptions options)
+    {
+        // Construct the data model.
+        var dataModel = new VectorStoreGenericDataModel<string>(storageModel.Id);
+
+        // Set Vector.
+        if (options?.IncludeVectors is true)
+        {
+            dataModel.Vectors.Add(this._propertyReader.FirstVectorPropertyName!, new ReadOnlyMemory<float>(storageModel.Values));
+        }
+
+        // Set Data.
+        if (storageModel.Metadata != null)
+        {
+            foreach (var dataProperty in this._propertyReader.DataProperties)
+            {
+                var propertyStorageName = this._propertyReader.GetStoragePropertyName(dataProperty.DataModelPropertyName);
```

Comment:
Does this throw if storage property name cannot be found or is null?

Reply:
DataModelPropertyName cannot be null. This is validated in the VectorStoreRecordProperty constructor.
Also, GetStoragePropertyName will always find something, since it's based on the same data properties list as we are using here and has a fallback mechanism for if no storage name is defined.  It just falls back to the data property name.